cmd/ui.go

```diff
@@ -0,0 +1,155 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2019 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package cmd
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/loadimpact/k6/core/local"
+	"github.com/loadimpact/k6/ui/pb"
+)
+
+// A writer that syncs writes with a mutex and, if the output is a TTY, clears before newlines.
+type consoleWriter struct {
+	Writer io.Writer
+	IsTTY  bool
+	Mutex  *sync.Mutex
+
+	// Used for flicker-free persistent objects like the progressbars
+	PersistentText func()
+}
+
+func (w *consoleWriter) Write(p []byte) (n int, err error) {
+	origLen := len(p)
+	if w.IsTTY {
+		//TODO: check how cross-platform this is...
+		p = bytes.ReplaceAll(p, []byte{'\n'}, []byte{'\x1b', '[', '0', 'K', '\n'})
+	}
+
+	w.Mutex.Lock()
+	n, err = w.Writer.Write(p)
+	if w.PersistentText != nil {
+		w.PersistentText()
+	}
+	w.Mutex.Unlock()
+
+	if err != nil && n < origLen {
+		return n, err
+	}
+	return origLen, err
+}
+
+func printBar(bar *pb.ProgressBar, rightText string) {
+	end := "\n"
+	if stdout.IsTTY {
+		//TODO: check for cross platform support
+		end = "\x1b[0K\r"
+	}
+	fprintf(stdout, "%s %s%s", bar.String(), rightText, end)
+}
+
+func renderMultipleBars(isTTY, goBack bool, pbs []*pb.ProgressBar) string {
+	lineEnd := "\n"
+	if isTTY {
+		//TODO: check for cross platform support
+		lineEnd = "\x1b[K\n" // erase till end of line
+	}
+
+	pbsCount := len(pbs)
+	result := make([]string, pbsCount+2)
+	result[0] = lineEnd // start with an empty line
+	for i, pb := range pbs {
+		result[i+1] = pb.String() + lineEnd
+	}
+	if isTTY && goBack {
+		// Go back to the beginning
+		//TODO: check for cross platform support
+		result[pbsCount+1] = fmt.Sprintf("\r\x1b[%dA", pbsCount+1)
+	} else {
+		result[pbsCount+1] = "\n"
+	}
+	return strings.Join(result, "")
+}
+
+//TODO: show other information here?
+//TODO: add a no-progress option that will disable these
+//TODO: don't use global variables...
+func showProgress(ctx context.Context, wg *sync.WaitGroup, conf Config, executor *local.Executor) {
+	defer wg.Done()
```

Comment:
I think this will be better without it ... wherever we need it (in the run.go)we can do it with 
```
go func(){
  defer wg.Done()
  showProgress(...)
}
```
Instead of having it know about waitgroups.

Reply:
done